Mathlib/LinearAlgebra/Eigenspace/Zero.lean

```diff
@@ -0,0 +1,197 @@
+/-
+Copyright (c) 2024 Johan Commelin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Johan Commelin
+-/
+import Mathlib.LinearAlgebra.Charpoly.ToMatrix
+import Mathlib.LinearAlgebra.Determinant
+import Mathlib.LinearAlgebra.Eigenspace.Minpoly
+import Mathlib.RingTheory.Artinian
+
+/-!
+# Results on the eigenvalue 0
+
+In this file we provide equivalent characterizations of properties related to the eigenvalue 0,
+such as being nilpotent, having determinant equal to 0, having a non-trivial kernel, etc...
+
+## Main results
+
+* `LinearMap.charpoly_nilpotent_tfae`:
+  equivalent characterizations of nilpotent endomorphisms
+* `LinearMap.hasEigenvalue_zero_tfae`:
+  equivalent characterizations of endomorphisms with eigenvalue 0
+* `LinearMap.not_hasEigenvalue_zero_tfae`:
+  endomorphisms without eigenvalue 0
+* `LinearMap.finrank_maximalGeneralizedEigenspace`:
+  the dimension of the maximal generalized eigenspace of an endomorphism
+  is the trailing degree of its characteristic polynomial
+
+-/
+
+variable {K M : Type*} [Field K] [AddCommGroup M] [Module K M] [Module.Finite K M]
+variable (φ : Module.End K M)
+
+open FiniteDimensional Module.Free Polynomial
+
+lemma IsNilpotent.charpoly_eq_X_pow_finrank (h : IsNilpotent φ) :
```

Comment:
I'm not convinced it's worth actually implementing this but FWIW, this result requires only `[CommRing K] [IsDomain K]` rather than `Field K` as long as you also add `[Module.Free K M]` (the proof needs no modification).

Reply:
Done